Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.cs

```diff
@@ -388,13 +391,49 @@ private void AttemptShoot(EntityUid user, EntityUid gunUid, GunComponent gun)
         if (!userImpulse || !TryComp<PhysicsComponent>(user, out var userPhysics))
             return;
 
+        RecoilKick(gun, (user, userPhysics), fromCoordinates, toCoordinates.Value);
+
         var shooterEv = new ShooterImpulseEvent();
         RaiseLocalEvent(user, ref shooterEv);
 
         if (shooterEv.Push || _gravity.IsWeightless(user, userPhysics))
             CauseImpulse(fromCoordinates, toCoordinates.Value, user, userPhysics);
     }
 
+    [Dependency] private readonly SharedStaminaSystem _stamina = default!;
+
+    private void RecoilKick(
+        GunComponent gun,
+        Entity<PhysicsComponent> user,
+        EntityCoordinates fromCoordinates,
+        EntityCoordinates toCoordinates
+    )
+    {
+        if (gun.RecoilKick is not { } kick ||
+            kick.Impulse <= 0.0f ||
+            !TryComp<RecoilKickSusceptibleComponent>(user, out var suscept))
+            return;
+
+        var attempt = new RecoilKickAttemptEvent();
+        RaiseLocalEvent(user, ref attempt);
+        var speed = kick.Impulse * attempt.Effectiveness * user.Comp.InvMass / suscept.MassFactor;
+
+        var fromMap = TransformSystem.ToMapCoordinates(fromCoordinates).Position;
+        var toMap = TransformSystem.ToMapCoordinates(toCoordinates).Position;
+        var shotDirection = (fromMap - toMap).Normalized();
+
+        ThrowingSystem.TryThrow(
+            user,
+            // Change the magnitude of `shotDirection` because the length _is_ actually used to determine flight time.
+            shotDirection * _Moffstation.Weapons.Ranged.Components.RecoilKick.FlyTime * speed,
+            speed,
+            doSpin: false
+        );
+
+        // deal stamina damage
+        _stamina.TakeStaminaDamage(user, speed * kick.StaminaMultiplier);
```

Comment:
Like I said, this should probably be separated out. It's also pretty untweaked right now, but it should be easy to tune.